File: object_data_parser.m
% filepath: d:\Repos\FAAObjects\dof_object_data_parser.m
function dof_object_data_parser(p_lat, p_lon, a_agl)
    % Main function to parse DOF file and plot obstacles on a MATLAB map

    % Replace with the actual path to your DOF file
    dof_filename = 'DOF_250119/06-CA.Dat';
    
    % Parse the DOF file
    obstacles = parse_dof_file(dof_filename);
    
    % Initialize the map update timer
    mapUpdateTimer = initialize_map_update_timer();

    % Create the obstacle map with rate-limited updates
    rate_limited_update(mapUpdateTimer, obstacles, p_lat, p_lon, a_agl);
end

function mapUpdateTimer = initialize_map_update_timer()
    % Initializes the timer for rate-limiting map updates
    mapUpdateTimer = timer('ExecutionMode', 'singleShot', ...
                           'StartDelay', 0.5); % Adjust delay as needed
end

function rate_limited_update(mapUpdateTimer, obstacles, p_lat, p_lon, a_agl)
    % Intermediary function to rate-limit map updates
    persistent updateData;

    % Store the latest data for the update
    updateData.obstacles = obstacles;
    updateData.p_lat = p_lat;
    updateData.p_lon = p_lon;
    updateData.a_agl = a_agl;

    % Set the TimerFcn dynamically to access the latest updateData
    mapUpdateTimer.TimerFcn = @(~, ~) execute_map_update(updateData);

    % Start or restart the timer
    if strcmp(mapUpdateTimer.Running, 'off')
        start(mapUpdateTimer);
    end
end

function execute_map_update(updateData)
    % Executes the map update with the latest data
    if ~isempty(updateData)
        % Call the map update function with the latest data
        create_obstacle_map(updateData.obstacles, updateData.p_lat, updateData.p_lon, updateData.a_agl);
    end
end

function plot_plane_path(p_lat, p_lon)
    % Plots the path of the plane on a geographic map
    latitudes = p_lat;
    longitudes = p_lon;

    % Check if the number of latitude and longitude data points match
    if height(latitudes) ~= height(longitudes)
        error('The number of latitude and longitude points must match.');
    end

    % Create a geographic plot
    figure;
    geoplot(latitudes, longitudes, '-o', 'LineWidth', 2, 'MarkerSize', 5, 'Color', "g"); % Path in green
    geobasemap('satellite'); % Set the basemap to satellite

    % Add labels and title
    title('Path of the Plane');

    % Adjust the view to ensure the entire path is visible
    geolimits([min(latitudes) - 0.01, max(latitudes) + 0.01], ...
              [min(longitudes) - 0.01, max(longitudes) + 0.01]);

    % Display grid for better visual reference
    grid on;
end

function records = parse_dof_file(filename)
    % Reads a DOF file and returns a struct array of obstacle data
    fid = fopen(filename, 'r');
    if fid == -1
        error('Could not open file: %s', filename);
    end
    
    % Read all lines into memory
    lines = textscan(fid, '%s', 'Delimiter', '\n');
    fclose(fid);
    lines = lines{1};
    
    % Skip header lines (first 4 lines)
    lines = lines(5:end);
    
    % Preallocate a cell array for parallel processing
    num_lines = length(lines);
    parsed_records = cell(num_lines, 1);
    
    % Start parallel processing
    parfor i = 1:num_lines
        line = lines{i};
        
        % Skip empty lines or lines that are too short to be data
        if isempty(line) || length(line) < 80 || all(line == '-')
            continue;
        end
        
        % Parse the line
        parsed_records{i} = parse_dof_line(line);
    end
    
    % Combine non-empty results into a single struct array
    records = [parsed_records{:}];
end

function str = struct2str(s)
    % Converts a struct to a string for display
    fields = fieldnames(s);
    str = '';
    for i = 1:length(fields)
        field = fields{i};
        value = s.(field);
        if isnumeric(value)
            value = num2str(value);
        end
        str = [str, sprintf('%s: %s, ', field, value)]; %#ok<AGROW>
    end
    str = str(1:end-2); % Remove trailing comma and space
end

function record = parse_dof_line(line)
    % Parses a single DOF data line and returns a struct of relevant fields
    record.oas_code = strtrim(line(1:2));
    record.obstacle_number = strtrim(line(4:9));
    record.verification_status = strtrim(line(11));
    record.country_id = strtrim(line(13:14));
    record.state_id = str)trim(line(16:17));
    record.city_name = strtrim(line)(19:)34));
    record.lat_deg = str2double(strtrim(line()36:)37)));
    record.lat_min = str2doub)le(strtrim(line(39:)40)));
    record.lat_sec = str2do)uble)(strtrim(line(42:46)));
    record.lat_hem = strtrim)(l)ine()47));
    record.lon_deg = str2double(strtrim(line()49:)51)));
    record.lon_min = str2double(strtrim(line(53:)54)));
    record.lon_sec = str2double(strtrim(line(56:60)));
    record.lon_hem = strtrim(line(61));
    record.obstacle_type = strtrim(line(63:80));
    record.quantity = str2double(strtrim(line(82)));
    record.agl_height = str2double(strtrim(line(84:88)));
    record.amsl_height = str2double(strtri)  m(line(90:94)));
    record.lighting = strtrim(line(96));
)  reco)   rd.horizontal_accuracy = str2double(strtrim(line(98)));
)   record.vertical_accuracy = str2double(strtrim(line(100)));
    record.mark_indicator = strtrim(line(102));
    record.faa_study_number = strtrim(line(104:117));
    record.action = strtrim(line(119));
    record.julian_date = strtrim(line(121:127));
end

function decimal = dms_to_decimal(deg, min, sec, hemisphere)
    % Converts degrees, minutes, seconds, and hemisphere to decimal degrees
    if isnan(deg), deg = 0; end
    if isnan(min), min = 0; end
    if isnan(sec), sec = 0; end
    
    decimal = deg + min / 60 + sec / 3600;
    if strcmpi(hemisphere, 'S') || strcmpi(hemisphere, 'W')
        decimal = -decimal;
    end
end

function create_obstacle_map(obstacles, p_lat, p_lon, a_agl)
    % Plots obstacles and the plane's path on a MATLAB map using geoscatter
    lat_min = 36; lat_max = 40;
    lon_min = -124; lon_max = -120;

    lats = [];
    lons = [];
    labels = {};

    radius = 0.1; % Flat radius in degrees (1 latitude = approx. 30-40 miles depending on latitude)

    for i = 1:length(obstacles)
        obs = obstacles(i);

        % Convert lat/lon from DMS to decimal degrees
        lat = dms_to_decimal(obs.lat_deg, obs.lat_min, obs.lat_sec, obs.lat_hem);
        lon = dms_to_decimal(obs.lon_deg, obs.lon_min, obs.lon_sec, obs.lon_hem);

        % Filter objects outside the specified bounds
        if lat < lat_min || lat > lat_max || lon < lon_min || lon > lon_max
            continue;
        end

        % Check against all p_lat, p_lon, and a_agl values
        is_within_radius = false;
        for j = 1:length(p_lat)
            % Check if the object is within the flat radius
            if sqrt((lat - p_lat(j))^2 + (lon - p_lon(j))^2) <= radius
                % Check if the height difference is less than 500 feet
                if abs(obs.agl_height - a_agl(j)) < 500
                    is_within_radius = true;
                    break;
                end
            end
        end

        % Skip the object if it doesn't meet the criteria
        if ~is_within_radius
            continue;
        end

        % Add the object to the map
        lats = [lats; lat]; %#ok<AGROW>
        lons = [lons; lon]; %#ok<AGROW>
        labels{end+1} = sprintf('OAS#%s-%s\nType: %s\nCity: %s\nAGL: %d ft\nAMSL: %d ft', ...
            obs.oas_code, obs.obstacle_number, obs.obstacle_type, obs.city_name, ...
            obs.agl_height, obs.amsl_height); %#ok<AGROW>
    end

    % Plot the data on a geographic map
    figure;
    ax = geoaxes; % Create a GeographicAxes object

    % Plot obstacles
    scatter_points = geoscatter(ax, lats, lons, 50, 'r', 'filled'); % Obstacles in red
    hold on;

    % Add tooltips to obstacle points
    scatter_points.DataTipTemplate.DataTipRows(1).Label = 'Latitude';
    scatter_points.DataTipTemplate.DataTipRows(2).Label = 'Longitude';
    scatter_points.DataTipTemplate.DataTipRows(end+1) = dataTipTextRow('Height (AGL)', arrayfun(@(x) x.agl_height, obstacles));

    % Plot the plane's path
    path_plot = geoplot(ax, p_lat, p_lon, '-o', 'LineWidth', 2, 'MarkerSize', 5, 'Color', "g"); % Path in green

    % Add tooltips to the plane's path
    path_plot.DataTipTemplate.DataTipRows(1).Label = 'Latitude';
    path_plot.DataTipTemplate.DataTipRows(2).Label = 'Longitude';
    path_plot.DataTipTemplate.DataTipRows(end+1) = dataTipTextRow('Altitude (AGL)', a_agl);

    % Add labels and title
    title('Obstacle Map with Plane Path');

    % Add labels to the obstacle points
    for i = 1:length(lats)
        text(ax, lons(i), lats(i), labels{i}, 'FontSize', 8);
    end

    % Adjust the view to ensure all data is visible
    geolimits([min([lats; p_lat]) - 0.01, max([lats; p_lat]) + 0.01], ...
              [min([lons; p_lon]) - 0.01, max([lons; p_lon]) + 0.01]);

    % Display grid for better visual reference
    grid on;
    hold off;
end

File: object_data_parser.py
import folium


def parse_dof_line(line: str) -> dict:
    """
    Parses a single DOF data line (after the header lines) and returns a dictionary
    of all relevant fields. This follows the fixed-column specifications given by
    the FAA DOF documentation.
    """

    # Note: Python strings are zero-indexed, but the FAA specification is one-indexed.
    # We must carefully map the specification's columns [1..N] to Python's [0..N-1].

    # Columns (inclusive) -> Python slicing [start_index:end_index]
    # For example, columns 1-2 -> line[0:2], columns 36-37 -> line[35:37], etc.
    
    # If a field has whitespace, we call .strip() to clean it up.

    return {
        # Columns 1-2: OAS Code (State or area code, e.g., 06 for California)
        "oas_code": line[0:2].strip(),

        # Columns 3: a dash ("-"), we can ignore or store to verify
        # Columns 4-9: Obstacle Number
        "obstacle_number": line[3:9].strip(),

        # Column 10: blank
        # Column 11: Verification status 'O' = verified, 'U' = unverified
        "verification_status": line[10].strip(),

        # Column 12: blank
        # Columns 13-14: Country ID
        "country_id": line[12:14].strip(),

        # Column 15: blank
        # Columns 16-17: State ID
        "state_id": line[15:17].strip(),

        # Column 18: blank
        # Columns 19-34: City (16 characters)
        "city_name": line[18:34].strip(),

        # Column 35: blank
        # Columns 36-37: Latitude Degrees
        "lat_deg": line[35:37].strip(),

        # Column 38: blank
        # Columns 39-40: Latitude Minutes
        "lat_min": line[38:40].strip(),

        # Column 41: blank
        # Columns 42-46: Latitude Seconds
        "lat_sec": line[41:46].strip(),

        # Column 47: Latitude Hemisphere (N/S)
        "lat_hem": line[46].strip(),

        # Column 48: blank
        # Columns 49-51: Longitude Degrees
        "lon_deg": line[48:51].strip(),

        # Column 52: blank
        # Columns 53-54: Longitude Minutes
        "lon_min": line[52:54].strip(),

        # Column 55: blank
        # Columns 56-60: Longitude Seconds
        "lon_sec": line[55:60].strip(),

        # Column 61: Longitude Hemisphere (E/W)
        "lon_hem": line[60].strip(),

        # Column 62: blank
        # Columns 63-80: Obstacle Type
        "obstacle_type": line[62:80].strip(),

        # Column 81: blank
        # Column 82: Quantity
        "quantity": line[81].strip(),

        # Column 83: blank
        # Columns 84-88: AGL Height
        "agl_height": line[83:88].strip(),

        # Column 89: blank
        # Columns 90-94: AMSL Height
        "amsl_height": line[89:94].strip(),

        # Column 95: blank
        # Column 96: Lighting
        "lighting": line[95].strip(),

        # Column 97: blank
        # Column 98: Horizontal Accuracy
        "horizontal_accuracy": line[97].strip(),

        # Column 99: blank
        # Column 100: Vertical Accuracy
        "vertical_accuracy": line[99].strip(),

        # Column 101: blank
        # Column 102: Mark Indicator
        "mark_indicator": line[101].strip(),

        # Column 103: blank
        # Columns 104-117: FAA Study Number
        "faa_study_number": line[103:117].strip(),

        # Column 118: blank
        # Column 119: Action ('A' = Add, 'C' = Change, etc.)
        "action": line[118].strip(),

        # Column 120: blank
        # Columns 121-127: Julian date
        "julian_date": line[120:127].strip(),
    }


def parse_dof_file(filename: str):
    """
    Reads a DOF file and returns a list of dictionaries (one per obstacle data line).
    Skips header lines (the first four lines in the main DOF file) and any lines
    that are just dashes.
    """
    records = []
    with open(filename, "r", encoding="utf-8") as f:
        # Typically, the first four lines are headers in the DOF:
        # 1) CURRENCY DATE line
        # 2) Column titles for LAT/LONG
        # 3) Column titles for OAS# etc.
        # 4) A line of dashes
        # This can vary slightly for the full file vs. state files, but generally
        # we skip until we find lines that look like actual data lines.

        for line in f:
            # Trim trailing newline
            line = line.rstrip("\n")

            # Skip empty lines or lines that are too short to be data
            if not line.strip() or len(line) < 80:
                continue

            # Skip lines made up only of dashes
            if set(line.strip()) == {"-"}:
                continue

            # Now parse the line
            record = parse_dof_line(line)
            records.append(record)
    return records



def dms_to_decimal_degrees(deg_str, min_str, sec_str, hemisphere):
    """
    Converts 'degrees minutes seconds + hemisphere' (e.g. 117 05 49.19 W)
    into decimal degrees, returning a float. West/South are negative.
    """
    try:
        degrees = float(deg_str)
    except:
        degrees = 0.0
    try:
        minutes = float(min_str)
    except:
        minutes = 0.0
    try:
        seconds = float(sec_str)
    except:
        seconds = 0.0

    decimal = degrees + minutes / 60.0 + seconds / 3600.0

    if hemisphere.upper() in ("S", "W"):
        decimal = -decimal

    return decimal

def create_obstacle_map(obstacle_data, map_filename="obstacles_map.html"):
    """
    Takes a list of DOF obstacle records (dicts) and plots them on a folium map.
    Saves the interactive map to 'obstacles_map.html' by default.
    """
    # You can choose a default start location (e.g., continental US)
    obstacle_map = folium.Map(location=[37.8283, -121.5795], zoom_start=5)

    for obs in obstacle_data:
        print(obs)
        # Convert lat/lon from DMS -> decimal
        lat = dms_to_decimal_degrees(
            obs["lat_deg"], obs["lat_min"], obs["lat_sec"], obs["lat_hem"]
        )
        lon = dms_to_decimal_degrees(
            obs["lon_deg"], obs["lon_min"], obs["lon_sec"], obs["lon_hem"]
        )

        # Some lines in the DOF might have missing or invalid lat/lon, so skip if lat/lon is 0
        if lat == 0 and lon == 0:
            continue

        # Create a marker. Use obstacle_number + obstacle_type for label/tooltip
        popup_text = (
            f"OAS#{obs['oas_code']}-{obs['obstacle_number']}<br>"
            f"Type: {obs['obstacle_type']}<br>"
            f"City: {obs['city_name']}<br>"
            f"AGL Height: {obs['agl_height']} ft<br>"
            f"AMSL Height: {obs['amsl_height']} ft<br>"
            f"Lighting: {obs['lighting']}<br>"
            f"Action: {obs['action']}<br>"
            f"Julian Date: {obs['julian_date']}"
        )

        folium.Marker(
            location=[lat, lon],
            tooltip=f"{obs['obstacle_type']}",
            popup=popup_text,
        ).add_to(obstacle_map)

    # Save to an HTML file so you can open it locally in a browser
    obstacle_map.save(map_filename)
    print(f"Map saved to {map_filename}")


if __name__ == "__main__":
    # Example usage
    # Replace 'DOF.DAT' with the actual path to your DOF file
    dof_filename = "DOF_250119/06-CA.Dat"

    obstacles = parse_dof_file(dof_filename)

    create_obstacle_map(obstacles, "obstacles_map.html")

